modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java

```diff
@@ -0,0 +1,124 @@
+import static org.apache.ignite.internal.utils.RebalanceUtil.partAssignmentsStableKey;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.ignite.configuration.schemas.table.TableConfiguration;
+import org.apache.ignite.internal.configuration.schema.ExtendedTableChange;
+import org.apache.ignite.internal.metastorage.MetaStorageManager;
+import org.apache.ignite.internal.metastorage.client.Entry;
+import org.apache.ignite.internal.metastorage.client.If;
+import org.apache.ignite.internal.raft.server.RaftGroupEventsListener;
+import org.apache.ignite.internal.util.ByteUtils;
+import org.apache.ignite.lang.ByteArray;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.raft.jraft.Status;
+import org.apache.ignite.raft.jraft.entity.PeerId;
+
+/**
+ * Listener for the raft group events, which must provide correct error handling of rebalance process
+ * and start new rebalance after the current one finished.
+ */
+public class RebalanceRaftGroupEventsListener implements RaftGroupEventsListener {
+
+    /** Meta storage manager. */
+    private final MetaStorageManager metaStorageMgr;
+
+    /** Table configuration instance. */
+    private final TableConfiguration tblConfiguration;
+
+    /** Unique partition id. */
+    private final String partId;
+
+    /** Partition number. */
+    private final int partNum;
+
+    /**
+     * Constructs new listener.
+     *
+     * @param metaStorageMgr Meta storage manager.
+     * @param tblConfiguration Table configuration.
+     * @param partId Partition id.
+     * @param partNum Partition number.
+     */
+    public RebalanceRaftGroupEventsListener(MetaStorageManager metaStorageMgr, TableConfiguration tblConfiguration, String partId,
+            int partNum) {
+        this.metaStorageMgr = metaStorageMgr;
+        this.tblConfiguration = tblConfiguration;
+        this.partId = partId;
+        this.partNum = partNum;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onLeaderElected() {
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onNewPeersConfigurationApplied(List<PeerId> peers) {
+        Map<ByteArray, Entry> keys = metaStorageMgr.getAll(
+                Set.of(partAssignmentsPlannedKey(partId), partAssignmentsPendingKey(partId))).join();
+
+        Entry plannedEntry = keys.get(partAssignmentsPlannedKey(partId));
+        Entry pendingEntry = keys.get(partAssignmentsPendingKey(partId));
+
+        tblConfiguration.change(ch -> {
+            List<List<ClusterNode>> assignments =
+                    (List<List<ClusterNode>>) ByteUtils.fromBytes(((ExtendedTableChange) ch).assignments());
+            assignments.set(partNum, ((List<ClusterNode>) ByteUtils.fromBytes(pendingEntry.value())));
+            ((ExtendedTableChange) ch).changeAssignments(ByteUtils.toBytes(assignments));
+        });
+
+        if (plannedEntry.value() != null) {
```

Comment:
Why do we need external plannedEntry null checking in addition to in-invoke
`notExists(partAssignmentsPlannedKey(partId)),` one? I'd rather say that invoke should be considered as an atomic operation and it's actually **is** because handling invoke command is linearized with raft.

All-in-all I suggest considering following operation instead

        /**
         * metastoreInvoke: \\ atomic
         *     partition.assignments.stable = appliedPeers
         *     if empty(partition.assignments.planned):
         *         partition.assignments.pending = empty
         *     else:
         *         partition.assignments.pending = partition.assignments.planned
         *         remove(partition.assignments.planned)
         *
         */
        if (!metaStorageMgr.invoke(If.iif(
                notExists(partAssignmentsPlannedKey(partId)),
                ops(
                        put(partAssignmentsStableKey(partId), appliedAssignments),
                        remove(partAssignmentsPendingKey(partId))
                )
                        .yield(true),
                ops(
                        put(partAssignmentsStableKey(partId), appliedAssignments),
                        put(partAssignmentsPendingKey(partId), get(partAssignmentsPlannedKey(partId))),
                        remove(partAssignmentsPlannedKey(partId))
                )
                        .yield(false))
        ).join().getAsBoolean()) {
            onNewPeersConfigurationApplied(peers);
        }

please pay attention that 

- instead of pre-readed plannedEntry of keys I used get operation that I suggest to add in addition to put, remove, noop ones.
- instead of pending keys resolved appliedAssignments are used.

Reply:
As was discussed, it'll be fixed later within separate ticket.